linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+//    // using an `else` because that confuses nvcc (11.2) into thinking there's not
+//    // a return statement
+//    if constexpr (sizeof...(n) == 0)
+//    {
+//       decltype(S{} * T{}) AB{};
+//       for (int i = 0; i < m; i++)
+//       {
+//          AB += A[i] * B[i];
+//       }
+//       return AB;
+//    }
+
+//    if constexpr (sizeof...(n) > 0)
+//    {
+//       constexpr int                     dimensions[] = {n...};
+//       tensor<decltype(S{} * T{}), n...> AB{};
+//       for (int i = 0; i < dimensions[0]; i++)
+//       {
+//          for (int j = 0; j < m; j++)
+//          {
+//             AB[i] = AB[i] + A[j] * B[j][i];
+//          }
+//       }
+//       return AB;
+//    }
+// }
+
+template <typename S, typename T, int m> MFEM_HOST_DEVICE
+auto dot(const tensor<S, m>& A, const tensor<T, m>& B) ->
+decltype(S {} * T{})
+{
+   decltype(S{} * T{}) AB{};
+   for (int i = 0; i < m; i++)
+   {
+      AB += A[i] * B[i];
+   }
+   return AB;
+}
+
+template <typename S, typename T, int m, int... n> MFEM_HOST_DEVICE
+auto dot(const tensor<S, m>& A, const tensor<T, m, n...>& B) ->
+tensor<decltype(S {} * T{}), n...>
+{
+   constexpr int dimensions[] = {n...};
+   tensor<decltype(S{} * T{}), n...> AB{};
+   for (int i = 0; i < dimensions[0]; i++)
+   {
+      for (int j = 0; j < m; j++)
+      {
+         AB[i] = AB[i] + A[j] * B[j][i];
+      }
+   }
+   return AB;
+}
+
+/**
+ * @overload
+ * @note vector . matrix . vector
+ */
+template <typename S, typename T, typename U, int m, int n> MFEM_HOST_DEVICE
+auto dot(const tensor<S, m>& u, const tensor<T, m, n>& A,
+         const tensor<U, n>& v) ->
+decltype(S {} * T{} * U{})
+{
+   decltype(S{} * T{} * U{}) uAv{};
+   for (int i = 0; i < m; i++)
+   {
+      for (int j = 0; j < n; j++)
+      {
+         uAv += u[i] * A[i][j] * v[j];
+      }
+   }
+   return uAv;
+}
+
+/**
+ * @brief double dot product, contracting over the two "middle" indices
+ * @tparam S the underlying type of the tensor (lefthand) argument
+ * @tparam T the underlying type of the tensor (righthand) argument
+ * @tparam m first dimension of A
+ * @tparam n second dimension of A
+ * @tparam p third dimension of A, first dimensions of B
+ * @tparam q fourth dimension of A, second dimensions of B
+ * @param[in] A The lefthand tensor
+ * @param[in] B The righthand tensor
+ */
+template <typename S, typename T, int m, int n, int p, int q> MFEM_HOST_DEVICE
+auto ddot(const tensor<S, m, n, p, q>& A, const tensor<T, p, q>& B) ->
+tensor<decltype(S {} * T{}), m, n>
+{
+   tensor<decltype(S{} * T{}), m, n> AB{};
+   for (int i = 0; i < m; i++)
+   {
+      for (int j = 0; j < n; j++)
+      {
+         for (int k = 0; k < p; k++)
+         {
+            for (int l = 0; l < q; l++)
+            {
+               AB[i][j] += A[i][j][k][l] * B[k][l];
+            }
+         }
+      }
+   }
+   return AB;
+}
+
+/**
+ * @overload
+ * @note 3rd-order-tensor : 2nd-order-tensor
```

Comment:
For clarity, add definition: "returns vector C, such that C_i = sum_jk A_ijk B_jk".